apps/ewallet/lib/ewallet/exchange/exchange.ex

```diff
@@ -0,0 +1,154 @@
+defmodule EWallet.Exchange do
+  @moduledoc """
+  Provides exchange functionalities.
+  """
+  alias EWallet.Exchange.Calculation
+  alias EWalletDB.{ExchangePair, Token}
+
+  # The private types for calculation parameters
+  @typep non_neg_or_nil() :: non_neg_integer() | nil
+
+  @doc """
+  Retrieves the exchange rate of the given token pair, adjusted for the `subunit_to_unit`
+  differences of the two tokens.
+
+  Returns `{:ok, rate, pair}` if the exchange pair is found.
+
+  If the returned pair is a reversed pair, the returned `rate` is already inverted.
+  """
+  @spec get_rate(%Token{}, %Token{}) :: {:ok, float(), ExchangePair.t()} | {:error, atom()}
+  def get_rate(from_token, to_token) do
+    subunit_scale = to_token.subunit_to_unit / from_token.subunit_to_unit
+
+    case ExchangePair.fetch_exchangable_pair(from_token, to_token) do
+      {:ok, pair, :direct} ->
+        # Direct pair. Return the rate directly.
+        {:ok, pair.rate * subunit_scale, pair}
+
+      {:ok, pair, :reversed} ->
+        # Reversed pair. Return the inverted rate.
+        {:ok, 1 / pair.rate * subunit_scale, pair}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  @doc """
+  Validates that the given `from_amount` and `to_amount` matches the exchange pair.
+
+  For same-token transactions, it returns `true` if `from_amount` and `to_amount` are equal,
+  otherwise returns `false`.
+
+  For cross-token transactions, it returns `true` if the amounts match the rate
+  of the exchange pair, otherwise returns `false`.
+  """
+  @spec validate(non_neg_or_nil(), %Token{}, non_neg_or_nil(), %Token{}) ::
+          {:ok, Calculation.t()} | {:error, atom()}
+
+  # Same-token: valid if `from_amount` and `to_amount` to be equal
+  def validate(from_amount, %{uuid: from_token} = token, to_amount, %{uuid: to_token})
```

Comment:
If I did this in one of my PR, you'd probably tell me to call it `from_token_uuid` instead of just `from_token`, just to clarify that we're manipulating an ID and not a token. :)

Reply:
Yeah I was compromising this with `mix format` trying to bump it to multiple lines and they look unreadable as hell to me 🤦‍♂️

Reply:
Wait... it doesn't do that anymore. What the hell.. 🤦‍♂️ 🤦‍♂️ 

Pushing the code now.